File: LeetCode/1020.number-of-enclaves.cpp
class Solution {
public:
    int numEnclaves(vector<vector<int>>& g) {
        int n = g.size();
        int m = g[0].size();

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (i == 0 || i == n - 1 || j == 0 || j == m - 1 && g[i][j] == 1) {
                    rec(i, j, g, n, m);
                }
            }
        }

        int cnt = 0;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (g[i][j] == 1) {
                    cnt++;
                }
            }
        }

        return cnt;
    }

    void rec(int i, int j, vector<vector<int>>& g, int n, int m) {
        if (i < 0 || j < 0 || i >= n || j >= m || g[i][j] == 0) {
            return;
        }

        g[i][j] = 0;

        rec(i + 1, j, g, n, m);
        rec(i, j + 1, g, n, m);
        rec(i - 1, j, g, n, m);
        rec(i, j - 1, g, n, m);
    }
};

File: LeetCode/1020.number-of-enclaves.py
#
# @lc app=leetcode id=1020 lang=python3
#
# [1020] Number of Enclaves
#

# @lc code=start
class Solution:
    def numEnclaves(self, grid: List[List[int]]) -> int:

        def dfs(i, j):
            if i < 0 or j < 0 or i >= len(grid) or j >= len(grid[0]):
                return
            if grid[i][j] == 0:
                return
            grid[i][j] = 0
            dfs(i+1, j)
            dfs(i-1, j)
            dfs(i, j+1)
            dfs(i, j-1)

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if (i == 0 or j == 0 or i == len(grid) - 1 or j == len(grid[0]) - 1) and grid[i][j] == 1:
                    dfs(i, j)

        ans = 0

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 1:
                    ans += 1

        return ans   
        
# @lc code=end

